conwayste/src/ui/pane.rs

```diff
@@ -16,56 +16,313 @@
 impl Pane {
     /// Specify the dimensional bounds of the Pane container
     pub fn new(dimensions: Rect) -> Self {
-        Pane {
+        let mut pane = Pane {
             id: None,
             z_index: std::usize::MAX,
-            dimensions: dimensions,
-            hover: false,
+            dimensions,
             floating: true,
             previous_pos: None,
             border: 1.0,
             bg_color: None,
+            focus_cycle: FocusCycle::new(CycleType::OpenEnded),
+            handler_data: context::HandlerData::new(),
+        };
+
+        // for each event type, define a handler of the appropriate type (mouse or keyboard)
+        for event_type in EventType::into_enum_iter() {
+            if event_type.is_mouse_event() {
+                let handler = |_obj: &mut dyn EmitEvent,
+                               uictx: &mut context::UIContext,
+                               evt: &context::Event|
+                 -> Result<Handled, Box<dyn Error>> {
+                    // let pane = obj.downcast_mut::<Pane>()?; // uncomment and rename _obj to obj above if we need a Pane
+
+                    for child_id in uictx.widget_view.children_ids() {
+                        let (widget_ref, mut subuictx) = uictx.derive(&child_id).unwrap(); // unwrap OK because 1) valid ID, 2) in view
+
+                        let point = &evt.point.unwrap(); // unwrap OK because a Click event always has a point
+                        if within_widget(&point, &widget_ref.rect()) {
+                            if let Some(emittable_ref) = widget_ref.as_emit_event() {
+                                emittable_ref.emit(evt, &mut subuictx)?;
+                                let pane_events = subuictx.collect_child_events();
+                                if pane_events.len() != 0 {
+                                    warn!("expected no mouse child events to be collected from Pane; got {:?}",
+                                        pane_events);
+                                }
+                                return Ok(Handled::Handled);
+                            } else {
+                                warn!(
+                                    "Widget at point of click ({:?}) does not implement EmitEvent: {:?}",
+                                    evt.point, widget_ref.id(),
+                                );
+                            }
+                        }
+                    }
+                    Ok(Handled::NotHandled)
+                };
+                pane.on(event_type, Box::new(handler)).unwrap(); // unwrap OK because we aren't calling from within a handler
+            } else if event_type.is_key_event() {
+                // unwrap OK because we aren't calling from within a handler
+                pane.on(event_type, Box::new(Pane::key_press_handler))
+                    .unwrap();
+            } else {
+                // nothing to do if this is not a key or a mouse event
```

Comment:
Should we log this?

Reply:
I don't think so. If we need to add a handler to all Panes we can do it manually. No need to warn about events the Pane doesn't know how to register handlers for.